src/Cli/dotnet/commands/dotnet-workload/install/WorkloadInstallCommand.cs

```diff
@@ -64,67 +66,97 @@ private void ValidateWorkloadIdsInput()
-            }
-            else
-            {
-                try
+                else if (_skipManifestUpdate && usedRollback)
                 {
-                    InstallWorkloads(
-                        _workloadIds.Select(id => new WorkloadId(id)),
-                        _skipManifestUpdate,
-                        _includePreviews,
-                        string.IsNullOrWhiteSpace(_fromCacheOption) ? null : new DirectoryPath(_fromCacheOption));
+                    throw new GracefulException(string.Format(LocalizableStrings.CannotCombineSkipManifestAndRollback,
+                        WorkloadInstallCommandParser.SkipManifestUpdateOption.Name, InstallingWorkloadCommandParser.FromRollbackFileOption.Name,
+                        WorkloadInstallCommandParser.SkipManifestUpdateOption.Name, InstallingWorkloadCommandParser.FromRollbackFileOption.Name), isUserError: true);
                 }
-                catch (Exception e)
+                else
                 {
-                    // Don't show entire stack trace
-                    throw new GracefulException(string.Format(LocalizableStrings.WorkloadInstallationFailed, e.Message), e, isUserError: false);
+                    recorder.Run(() =>
+                    {
+                        try
+                        {
+                            InstallWorkloads(
+                                _workloadIds.Select(id => new WorkloadId(id)),
+                                _skipManifestUpdate,
+                                _includePreviews,
+                                string.IsNullOrWhiteSpace(_fromCacheOption) ? null : new DirectoryPath(_fromCacheOption),
+                                manifestRollbacks);
+                        }
+                        catch (Exception e)
+                        {
+                            // Don't show entire stack trace
+                            throw new GracefulException(string.Format(LocalizableStrings.WorkloadInstallationFailed, e.Message), e, isUserError: false);
+                        }
+                    });
                 }
             }
-
-            _workloadInstaller.Shutdown();
+            finally
+            {
+                _workloadInstaller.Shutdown();
+            }
+            
             return _workloadInstaller.ExitCode;
         }
 
-        public void InstallWorkloads(IEnumerable<WorkloadId> workloadIds, bool skipManifestUpdate = false, bool includePreviews = false, DirectoryPath? offlineCache = null)
+        public void InstallWorkloads(
+            IEnumerable<WorkloadId> workloadIds,
+            bool skipManifestUpdate = false,
+            bool includePreviews = false,
+            DirectoryPath? offlineCache = null,
+            IEnumerable<(ManifestId, ManifestVersion, SdkFeatureBand)> rollbackContents = null)
```

Comment:
Is this parameter used at all?  It looks like it's not.

If it is, similar to other changes I've suggested, I think the type of this parameter can be `WorkloadRollbackInfo`.

Reply:
I restructured how this works to pass the WorkloadHistoryRecorder forward so we don't have to calculate the rollback info early and lug it around. It looks a lot cleaner this way (to me, at least).